crates/factor-outbound-networking/src/runtime_config/spin.rs

```diff
@@ -0,0 +1,273 @@
+            ca_roots_file,
+            client_cert_file,
+            client_private_key_file,
+        } = toml_config;
+        ensure!(
+            !component_ids.is_empty(),
+            "[[client_tls]] 'component_ids' list may not be empty"
+        );
+        ensure!(
+            !hosts.is_empty(),
+            "[[client_tls]] 'hosts' list may not be empty"
+        );
+
+        let components = component_ids.into_iter().map(Into::into).collect();
+
+        let hosts = hosts
+            .iter()
+            .map(|host| {
+                let host: &str = host;
+                host.parse()
+                    .map_err(|err| anyhow::anyhow!("invalid host {host:?}: {err:?}"))
+            })
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let use_webpki_roots = if let Some(ca_use_webpki_roots) = ca_use_webpki_roots {
+            ca_use_webpki_roots
+        } else {
+            // Use webpki roots by default *unless* explicit roots were given
+            ca_roots_file.is_none()
+        };
+
+        let root_certificates = ca_roots_file
+            .map(|path| self.load_certs(path))
+            .transpose()?
+            .unwrap_or_default();
+
+        let client_cert = match (client_cert_file, client_private_key_file) {
+            (Some(cert_path), Some(key_path)) => Some(super::ClientCertConfig {
+                cert_chain: self.load_certs(cert_path)?,
+                key_der: self.load_key(key_path)?,
+            }),
+            (None, None) => None,
+            (Some(_), None) => bail!("client_cert_file specified without client_private_key_file"),
+            (None, Some(_)) => bail!("client_private_key_file specified without client_cert_file"),
+        };
+
+        Ok(TlsConfig {
+            components,
+            hosts,
+            root_certificates,
+            use_webpki_roots,
+            client_cert,
+        })
+    }
+
+    // Parse certs from the provided file
+    fn load_certs(
+        &self,
+        path: impl AsRef<Path>,
+    ) -> io::Result<Vec<rustls_pki_types::CertificateDer<'static>>> {
+        let path = self.runtime_config_dir.join(path);
+        rustls_pemfile::certs(&mut io::BufReader::new(fs::File::open(path).map_err(
+            |err| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("failed to read cert file {:?}", err),
+                )
+            },
+        )?))
+        .collect::<io::Result<Vec<rustls_pki_types::CertificateDer<'static>>>>()
+    }
+
+    // Parse a private key from the provided file
+    fn load_key(
+        &self,
+        path: impl AsRef<Path>,
+    ) -> anyhow::Result<rustls_pki_types::PrivateKeyDer<'static>> {
+        let path = self.runtime_config_dir.join(path);
+        rustls_pemfile::private_key(&mut io::BufReader::new(
+            fs::File::open(path).context("loading private key")?,
+        ))
+        .map_err(|_| anyhow::anyhow!("invalid input"))
```

Comment:
The error here can probably be improved.

```suggestion
        .with_context(|_| format!("failed to get private key from '{}'", path.display()))
```